### PR TITLE
adopt nodeTemplate in MachineClasses

### PR DESCRIPTION
**Pull request by AxiomSamarth**
**How to categorize this PR?**

/area auto-scaling
/kind enhancement
/platform aws

**What this PR does / why we need it**:
- This PR adopts `nodeTemplate` field from the `worker` resource while creating the `MachineClasses` as needed by Cluster Autoscaler for a generic scale from zero.
- Also to achieve the scale to/from zero for the scaleMP use case, this PR also enhances the providerConfig as explained in the comment [here](https://github.com/gardener/gardener/issues/4948#issuecomment-997815132)

**Which issue(s) this PR fixes**:
Fixes [Autoscaler/#27](https://github.com/gardener/autoscaler/issues/27)

**Special notes for your reviewer**:

**Release note**:

```feature user
NONE
```

### CODE REVIEW

**Comment by AxiomSamarth** [comment]
@dkistner @himanshu-kun I have adopted the `providerConfig` changes in the commit [here](https://github.com/gardener/gardener-extension-provider-aws/pull/461/commits/a34f23dbb5d28e82a7f84780a5a901f23ca685d5) as per [this](https://github.com/gardener/gardener/issues/4948#issuecomment-997815132) discussion comment. Could you provide some early review comments?

File: pkg/apis/aws/types_worker.go
Comment (himanshu-kun) [inline pkg/apis/aws/types_worker.go:43]: This can be removed as we can use the NodeTemplate struct defined in g/g

File: pkg/apis/aws/v1alpha1/types_worker.go
Comment (himanshu-kun) [inline pkg/apis/aws/v1alpha1/types_worker.go:49]: same here

File: pkg/controller/worker/machines.go
Comment (himanshu-kun) [inline pkg/controller/worker/machines.go:180]: This can be better written as if..elseif
```
if workerConfigNodeTemplate is provided{
                  // use it
} else if poolNodeTemplate is provided{
                  // use it
}
```

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
Thanks for the quick PR @AxiomSamarth 
Some code refactoring changes requested.

File: pkg/apis/aws/types_worker.go
Comment (AxiomSamarth) [reply pkg/apis/aws/types_worker.go:43]: Good catch.

File: pkg/apis/aws/validation/worker.go
Comment (himanshu-kun) [reply pkg/apis/aws/validation/worker.go:92]: Also please add validation logic that `negative` value is not provided for `cpu` , `gpu` or `memory` ..you can use the fn `ValidateResourceQuantity()` in g/g for that.

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
some validation related changes asked

**Review by himanshu-kun — COMMENTED** [review]
Thanks it look great now! 
will approve once dependencies are vendored and tests are passing.

**Comment by AxiomSamarth** [comment]
> Thanks it look great now!
> will approve once dependencies are vendored and tests are passing.

The [dependencies](https://github.com/AxiomSamarth/gardener-extension-provider-aws/blob/48b0b294a1b305cc1734b986b365cb24a5702c58/go.mod#L12-L13) are already vendored (if you are talking about g/g v1.38 & MCM v0.42.0) & unit tests are passing too. 

Let's wait for the approval of this approach as there is no concrete approval for it yet. Once done, I will make this PR from draft to ready for review.

**Comment by himanshu-kun** [comment]
@AxiomSamarth Can you also update the doc to enhance it with nodeTemplate being part of `WorkerConfig`?

File: pkg/controller/worker/machines.go
Comment (dkistner) [inline pkg/controller/worker/machines.go:185]: Just to confirm: I assume we give the configuration in the providerConfig priority over the pool config due to the scenarios where the user provide the values on its own e.g. scaleMP?

File: pkg/apis/aws/validation/worker.go
Comment (dkistner) [inline pkg/apis/aws/validation/worker.go:94]:
```suggestion
		for _, capacityAttribute := range []corev1.ResourceName{"cpu", "gpu", "memory"} {
			value, ok := nodeTemplate.Capacity[capacityAttribute]
			if !ok {
				allErrs = append(allErrs, field.Required(fldPath.Child("nodeTemplate").Child("capacity"), fmt.Sprintf("%s is a mandatory field", capacityAttribute)))
				continue
			}
			allErrs = append(allErrs, validateResourceQuantityValue(capacityAttribute, value, fldPath.Child("nodeTemplate").Child("capacity").Child(string(capacityAttribute)))...)
		}
```

Nit

File: pkg/controller/worker/machines.go
Comment (AxiomSamarth) [reply pkg/controller/worker/machines.go:185]: That's right, @dkistner.  The configuration in the `providerConfig` is prioritised over the pool config due to the scenarios where the user provide the values on its own e.g. scaleMP

Comment (himanshu-kun) [reply pkg/controller/worker/machines.go:185]: Yes you are correct.

**Review by dkistner — APPROVED** [review]
/lgtm

**Pull request merged**
